File: Explorer/MicrosoftOrganizationChartReader/Records/ShadowRecord.cs
using System.ComponentModel;
using System.Drawing;
using System.IO;

namespace OrganizationChartReader
{
    public class ShadowRecord
    {
        public ushort Size { get; set; }

        public ShadowType Type { get; set; }

        public short Height { get; set; }

        public short Width { get; set; }

        public Color FillColor { get; set; }

        public Color Color { get; set; }

        public ShadowRecord(BinaryReader reader)
        {
            Size = reader.ReadUInt16();
            Type = (ShadowType)reader.ReadUInt16();
            Height = reader.ReadInt16();
            Width = reader.ReadInt16();
            FillColor = reader.ReadOrganizationChartColor();
            Color = reader.ReadOrganizationChartColor();
        }

        public enum ShadowType
        {
            None,
            SideAndBottom = 0x01,
            SideAndBottomBevel = 0x02,
            TopOrBottom = 0x03,
            LeftBottomRight = 0x04
        }
    }
}

File: Sources/MicrosoftOrganizationChartReader/Records/ShadowRecord.swift
//
//  ShadowInfoRecord.swift
//
//
//  Created by Hugh Bellamy on 20/01/2021.
//

import DataStream

public struct ShadowRecord {
    public let size: UInt16
    public let type: ShadowType
    public let height: UInt16
    public let width: UInt16
    public let fillColor: (red: UInt8, blue: UInt8, green: UInt8, unused: UInt8)
    public let color: (red: UInt8, blue: UInt8, green: UInt8, unused: UInt8)

    public init(dataStream: inout DataStream) throws {
        self.size = try dataStream.read(endianess: .littleEndian)
        self.type = try ShadowType(dataStream: &dataStream)
        self.height = try dataStream.read(endianess: .littleEndian)
        self.width = try dataStream.read(endianess: .littleEndian)
        self.fillColor = (
            red: try dataStream.read(endianess: .littleEndian),
            blue: try dataStream.read(endianess: .littleEndian),
            green: try dataStream.read(endianess: .littleEndian),
            unused: try dataStream.read(endianess: .littleEndian)
        )
        self.color = (
            red: try dataStream.read(endianess: .littleEndian),
            blue: try dataStream.read(endianess: .littleEndian),
            green: try dataStream.read(endianess: .littleEndian),
            unused: try dataStream.read(endianess: .littleEndian)
        )
    }
    
    public enum ShadowType: UInt16, DataStreamCreatable {
        case none = 0x00
        case sideAndBottom = 0x01 // Left/Bottom and Right/Bottom
        case sideAndBottomBevel = 0x02 // Left/Bottom/Bevel and Right/Bottom/Bevel
        case topOrBottom = 0x03
        case leftBottomRight = 0x04
        case topBottomLeftRightBevel = 0x05 // Not documented
        case topBottomLeftRight = 0x06 // Not documented
        case topRight = 0x07 // Not documented
        case leftBottom = 0x08 // Not documented
        case inset = 0x09 // Not documented
    }
}
